contracts/WooPPV2.sol

```diff
@@ -0,0 +1,384 @@
+* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+* copies of the Software, and to permit persons to whom the Software is
+* furnished to do so, subject to the following conditions:
+*
+* The above copyright notice and this permission notice shall be included in all
+* copies or substantial portions of the Software.
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+import './libraries/InitializableOwnable.sol';
+import './libraries/DecimalMath.sol';
+import './interfaces/IWooracleV2.sol';
+import './interfaces/IWooPPV2.sol';
+import './interfaces/IWooFeeManager.sol';
+import './interfaces/IWooGuardian.sol';
+import './interfaces/AggregatorV3Interface.sol';
+
+import '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+import '@openzeppelin/contracts/token/ERC20/ERC20.sol';
+import '@openzeppelin/contracts/token/ERC20/SafeERC20.sol';
+import '@openzeppelin/contracts/utils/ReentrancyGuard.sol';
+import '@openzeppelin/contracts/utils/Pausable.sol';
+import '@openzeppelin/contracts/math/SafeMath.sol';
+import '@uniswap/lib/contracts/libraries/TransferHelper.sol';
+
+/// @title Woo private pool for swaping.
+/// @notice the implementation class for interface IWooPPV2, mainly for query and swap tokens.
+contract WooPPV2 is InitializableOwnable, ReentrancyGuard, Pausable, IWooPPV2 {
+    /* ----- Type declarations ----- */
+
+    using SafeMath for uint256;
+    using DecimalMath for uint256;
+    using SafeERC20 for IERC20;
+
+    /* ----- State variables ----- */
+
+    mapping(address => bool) public isAdmin;
+    mapping(address => uint256) public reserves;
+
+    /// @inheritdoc IWooPPV2
+    address public override quoteToken;
+
+    IWooracleV2 public wooracle;
+
+    IWooFeeManager public feeManager;
+
+    uint256 public feeRate;
+    uint256 public unclaimedFee;
+
+    /* ----- Modifiers ----- */
+
+    modifier onlyAdmin() {
+        require(msg.sender == _OWNER_ || isAdmin[msg.sender], 'WooPPV2: !admin');
+        _;
+    }
+
+    constructor() public {
+        initOwner(msg.sender);
+    }
+
+    function init(
+        address _quoteToken,
+        address _wooracle,
+        address _feeManager
+    ) external onlyOwner {
+        require(_quoteToken == address(0), 'WooPPV2: INIT_INVALID');
+
+        quoteToken = _quoteToken;
+        wooracle = IWooracleV2(_wooracle);
+        feeManager = IWooFeeManager(_feeManager);
+        require(feeManager.quoteToken() == _quoteToken, 'WooPPV2: !feeManager');
+
+        feeRate = 2500000000000; // TODO
```

Comment:
According to the last meeting, feeRate should still using mapping instead of a variable?